vm/devices/storage/disklayer_sqlite/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+        })
+        .await
+        .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        Ok(())
+    }
+
+    fn unmap_behavior(&self) -> UnmapBehavior {
+        // This layer zeroes if the lower layer is zero, but otherwise does
+        // nothing, so we must report unspecified.
+        UnmapBehavior::Unspecified
+    }
+
+    fn optimal_unmap_sectors(&self) -> u32 {
+        1
+    }
+}
+
+impl WriteNoOverwrite for SqliteDiskLayer {
+    async fn write_no_overwrite(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+    ) -> Result<(), DiskError> {
+        self.write_maybe_overwrite(buffers, sector, false).await
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+enum SectorKind {
+    AllZero,
+    AllOne,
+    Data(Vec<u8>),
+}
+
+// FUTURE: pass RequestBuffers into this function directly, and avoid the alloc
+// overhead.
+fn read_sectors(
+    conn: Arc<Mutex<Connection>>,
+    sector_size: u32,
+    start_sector: u64,
+    end_sector: u64,
+) -> Result<Vec<(u64, SectorKind)>, rusqlite::Error> {
+    let conn = conn.lock();
+
+    let mut select_stmt = conn.prepare_cached(
+        "SELECT sector, data
+        FROM sectors
+        WHERE sector >= ? AND sector < ?
+        ORDER BY sector ASC",
+    )?;
+    let mut rows = select_stmt.query(rusqlite::params![start_sector, end_sector])?;
+
+    // reuse the same blob handle across all reads.
+    //
+    // this handle isn't easy to persist across multiple calls to
+    // `read_sectors`, due to the lifetime restrictions imposed by using a
+    // `Arc<Mutex<Connection>>`.
+    let Some(first_row) = rows.next()? else {
+        return Ok(Vec::new());
+    };
+    let mut sector: u64 = first_row.get(0)?;
+    let mut blob = conn.blob_open(
+        rusqlite::DatabaseName::Main,
+        "sectors",
+        "data",
+        sector as i64,
+        true,
+    )?;
+
+    let mut buf = vec![0; sector_size as usize];
+    let mut res = Vec::new();
+    loop {
+        let data = match blob.len() {
+            0 => SectorKind::AllZero,
+            1 => SectorKind::AllOne,
+            len => {
+                if len != sector_size as usize {
+                    return Err(rusqlite::Error::BlobSizeError);
+                }
+                blob.read_at(&mut buf, 0)?;
+                SectorKind::Data(buf.clone())
+            }
+        };
+
+        res.push((sector, data));
+
+        let Some(next_row) = rows.next()? else {
+            break;
+        };
+
+        sector = next_row.get(0)?;
+        blob.reopen(sector as i64)?;
+    }
+
+    Ok(res)
+}
+
+// FUTURE: pass RequestBuffers into this function directly, and avoid the alloc
+// overhead.
+fn write_sectors(
+    conn: Arc<Mutex<Connection>>,
+    sector_size: u32,
+    mut sector: u64,
+    buf: Vec<u8>,
+    overwrite: bool,
+) -> Result<(), rusqlite::Error> {
+    let mut conn = conn.lock();
```

Comment:
Should we use an async `Mutex` that we acquire via `lock_owned`, before calling `unblock`? That way, we wouldn't spin up a thread per IO just to wait for the connection lock.

Reply:
I think this ties into the connection pool suggestion? If so, tying this back to https://github.com/microsoft/openvmm/pull/380#discussion_r1878557865

Reply:
Oh, I see - even though we're burning threads via unblock, the async mutex will at least prevent that from burning threads on the caller side?

Reply:
P.S: what's the current blessed async mutex implementation in-tree?

Reply:
The idea is that since we know there will be long wait times (since we're doing IO under the lock), do the wait for the connection on the async side, then pass it over to the sync side only once it's available. The wait is cheaper on the async side.

You can use the one from `futures`. It has a handy `lock_owned` method on `Arc<Mutex<T>>` so that you don't have a lifetime problem when handing off to `unblock`.